validation-test/Sema/type_checker_perf/slow/issue43369.swift

```diff
@@ -0,0 +1,24 @@
+// RUN: %target-typecheck-verify-swift -solver-expression-time-threshold=1
+// REQUIRES: tools-release,no_asan
+
+// FIXME: Removing the time threshold improves diagnostics, but takes ~10s.
+// https://github.com/apple/swift/issues/43369
+
+struct A {}
+struct B {}
+
+func - (lhs: A, rhs: A) -> B {}
+func / (lhs: B, rhs: Double) -> B {}
+func + (lhs: B, rhs: B) -> B {}
+func - (lhs: B, rhs: B) -> B {}
+
+do {
+  let (a0, a1, a2, a3) = (A(), A(), A(), A())
+  let (t0, t1, t2) = (0, 1, 2)
+
+  // "binary operator '/' cannot be applied to operands of type 'B' and 'Int'"
+  // w/o threshold.
+  //
+  // expected-error@+1 3 {{cannot convert value of type 'B' to expected argument type 'Int'}}
```

Comment:
Could it be that it gets solved in time, but then the possibly orthogonal computation responsible for the more accurate diagnostic when the threshold is removed takes unexpectedly long?

Reply:
Could you please add `isTooComplex` check to the `ConstraintSystem::salvage()` and check whether that would make a difference here?

Reply:
You'd actually want to do that all the way before even the debug message since we don't really want to try and resolve if the time is expired, it's just a waste of time.

Reply:
> You'd actually want to do that all the way before even the debug message

How would you go about calling `isTooComplex` without a solution at hand?

Reply:
I know what's going on. The reason we enter "diagnostics" mode non-deterministically with a time threshold of 1s is that "performance" mode completes in just under or, much more rarely, over a second—it is simply a matter of time variation. 

> I tried to run the example with `-Xfrontend -debug-constraints` and it consistently fails with "too complex"

`-Xfrontend -debug-constraints` implies a certain amount of overhead—enough for "performance" mode to consistently outrun the 1s mark and [conclude](https://github.com/apple/swift/blob/7f9e39c22c789037bc6cce0bad58b1e76aaee89d/lib/Sema/CSSolver.cpp#L1386) the expression was too complex.

> Another interesting question is why doesn't `isTooComplex` produce `true` while solving in "diagnostic" mode

It would have produced `true` if it was reached. In our case, we hit a threshold during [`solveImpl`](https://github.com/apple/swift/blob/7f9e39c22c789037bc6cce0bad58b1e76aaee89d/lib/Sema/ConstraintSystem.cpp#L3834) and then [return an ambiguous result](https://github.com/apple/swift/blob/7f9e39c22c789037bc6cce0bad58b1e76aaee89d/lib/Sema/ConstraintSystem.cpp#L3839) before checking for `isTooComplex`. 
___
It appears that adding a `isTooComplex` check after `solveImpl` was the right fix, the question is whether the other, later check should be kept in place.

Reply:
I understand the race between normal/diagnostic modes but what bothers me is we'd produce a conversion error with 1 sec. threshold and without it just keep going until the solver hits default time/memory threshold and fails with "too complex".  `BindingSet::take` has `isTooComplex` check which means that every disjunction choice and variable binding should have triggered that check and failed so none of the "diagnostic" paths should have been successful yet somehow `solveImpl` ends up with solutions...

Reply:
> before hitting it

The fact that "too complex" has been hit should produce "too complex" disregard any solutions, that's probably what is not happening... Also producing 1800 solutions sounds like a separate bug to me.

Reply:
You mean that hitting the threshold in `BindingSet::take` should not only abort the step but also cause all previously found solutions to be dropped before exiting `solveImpl`?

Reply:
Maybe not dropped but disregarded instead, we need to add isTooComplex check before filtering in salvage.

Reply:
I mean all the way after `solveImpl` is done so we don't even attempt to diagnose any ambiguities.

Reply:
> we need to add isTooComplex check before filtering in salvage

By "filtering" are you referring to `findBestSolution`?

Reply:
Yeah, but see my previous comment.

Reply:
> I mean all the way after `solveImpl` is done so we don't even attempt to diagnose any ambiguities

Wait, is that not what I did yesterday, when you asked me to add an `isTooComplex` check to `salvage`?:

> Alright, hoisting the [isTooComplex check](https://github.com/apple/swift/blob/380e370157655a102e6cf1abd46b74d575f2ecce/lib/Sema/ConstraintSystem.cpp#L3876) in `ConstraintSystem::salvage()` to right after the `solveImpl` call seems to do the job.

> It appears that adding a `isTooComplex` check after `solveImpl` was the right fix, the question is whether the other, later check should be kept in place.

Reply:
> Wait, is that not what I did yesterday, when you asked me to add an isTooComplex check to salvage?

Kind of yes, but it was an overkill to put it before solveImpl because steps check that already, the problem actually was with solutions which I didn't realize could happen...

> It appears that adding a isTooComplex check after solveImpl was the right fix, the question is whether the other, later check should be kept in place.

That's a good question, I think what we want to do is to move `isTooComplex` into `findBestSolution` if it doesn't have one already and remove it from `salvage()` that way "too complex" covers both - post-solving diagnostic emission and filtering.

Reply:
> That's a good question, I think what we want to do is to move `isTooComplex` into `findBestSolution` if it doesn't have one already and remove it from `salvage()` that way "too complex" covers both - post-solving diagnostic emission and filtering

In our case we return prior to `findBestSolution` (on the `diagnoseAmbiguityWithFixes` check), so this will not work unless we also have a `isTooComplex` check right after the `solveImpl` call. Besides, `findBestSolution` does not return a `SolutionResult`.

Reply:
What I'm proposing is following:

- add `isTooComplex` check right after `solveImpl` (`solveImpl` itself is already covered because steps check `isTooComplex` and fail) that would cover the race condition between solve and diagnostics; 
- Verify whether `findBestSolution` has `isTooComplex` internally and if not add it there to make sure that solver doesn't spend too much time trying to filter solutions; 
- remove existing check from `salvage()`

Reply:
Ah, sorry, I thought your latest suggestion was an alternative. SGTM.

Reply:
Done! For reference, I tried adapting the example for a scale test, but the memory threshold gets hit as early as N=2, which makes the curve appear sublinear.